Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [2]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv('../heart_pandas.csv')

In [4]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

> 1/1 баллов

In [6]:
data['sex'].value_counts()

male      207
female     96
Name: sex, dtype: int64

In [ ]:
## Ваш ответ:male-207, female-96.

 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

> 1/1 баллов

In [7]:
data['sex'].value_counts(normalize=True)

male      0.683168
female    0.316832
Name: sex, dtype: float64

In [33]:
## Ваш ответ: 
print('male',0.683168*100,'%')

male 68.3168 %


**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

> 1/1 баллов

In [34]:
pd.crosstab(data['target'], data['sex'])

sex     female  male
target              
0           24   114
1           72    93

In [18]:
## Ваш ответ: male-93,female - 72.

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

> 1/1 баллов

In [49]:
pd.crosstab(data['sex'], data['target'], normalize=True)

target         0         1
sex                       
female  0.079208  0.237624
male    0.376238  0.306931

In [19]:
## Ваш ответ:37.6238

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

нужно получить это чило, а не смотреть глазами. Переделай и пришли в личку

> 1/1 баллов

In [48]:
data.sort_values(by=['target', 'age'],
        ascending=[False, True]).head()

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
72    29    male    typical angina                     130          204   
58    34    male  non-anginal pain                     118          182   
125   34  female    typical angina                     118          210   
65    35  female    typical angina                     138          183   
157   35    male    typical angina                     122          192   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
72   lower than 120mg/ml                 normal                      202   
58   lower than 120mg/ml                 normal                      174   
125  lower than 120mg/ml  ST-T wave abnormality                      192   
65   lower than 120mg/ml  ST-T wave abnormality                      182   
157  lower than 120mg/ml  ST-T wave abnormality                      174   

    exercise_induced_angina  st_depression st_slope  num_major_vessels  \
72                       no            0.0     flat                  0   
58                       no            0.0     flat                  0   
125                      no            0.7     flat                  0   
65                       no            1.4     flat                  0   
157                      no            0.0     flat                  0   

      thalassemia  target  
72   fixed defect       1  
58   fixed defect       1  
125  fixed defect       1  
65   fixed defect       1  
157  fixed defect       1

In [ ]:
## Ваш ответ:29

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

> 1/1 баллов

In [26]:
data.sort_values(by=['target', 'age'],
        ascending=[True, False]).head()

age   sex  chest_pain_type  resting_blood_pressure  cholesterol  \
238   77  male   typical angina                     125          304   
225   70  male   typical angina                     145          174   
234   70  male   typical angina                     130          322   
240   70  male  atypical angina                     160          269   
249   69  male  atypical angina                     140          254   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
238  lower than 120mg/ml                 normal                      162   
225  lower than 120mg/ml  ST-T wave abnormality                      125   
234  lower than 120mg/ml                 normal                      109   
240  lower than 120mg/ml  ST-T wave abnormality                      112   
249  lower than 120mg/ml                 normal                      146   

    exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
238                     yes            0.0       flat                  3   
225                     yes            2.6  upsloping                  0   
234                      no            2.4  upsloping                  3   
240                     yes            2.9  upsloping                  1   
249                      no            2.0  upsloping                  3   

           thalassemia  target  
238       fixed defect       0  
225  reversable defect       0  
234       fixed defect       0  
240  reversable defect       0  
249  reversable defect       0

In [ ]:
## Ваш ответ:77

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

> 1/1 баллов

In [27]:
data.sort_values(by=['target','sex', 'age'],
        ascending=[False,True, True]).head()

age     sex  chest_pain_type  resting_blood_pressure  cholesterol  \
125   34  female   typical angina                     118          210   
65    35  female   typical angina                     138          183   
115   37  female  atypical angina                     120          215   
124   39  female  atypical angina                      94          199   
154   39  female  atypical angina                     138          220   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
125  lower than 120mg/ml  ST-T wave abnormality                      192   
65   lower than 120mg/ml  ST-T wave abnormality                      182   
115  lower than 120mg/ml  ST-T wave abnormality                      170   
124  lower than 120mg/ml  ST-T wave abnormality                      179   
154  lower than 120mg/ml  ST-T wave abnormality                      152   

    exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
125                      no            0.7       flat                  0   
65                       no            1.4       flat                  0   
115                      no            0.0       flat                  0   
124                      no            0.0       flat                  0   
154                      no            0.0  upsloping                  0   

      thalassemia  target  
125  fixed defect       1  
65   fixed defect       1  
115  fixed defect       1  
124  fixed defect       1  
154  fixed defect       1

In [22]:
## Ваш ответ:34

**Какой средний возраст женщин?**

> 1/1 баллов

In [28]:
data[data['sex'] == 'female']['age'].mean()

55.677083333333336

In [29]:
## Ваш ответ:55.677083333333336

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

> 1/1 баллов

In [71]:
data.groupby(['target'])['age'].describe(percentiles=[])

count       mean       std   min   50%   max
target                                              
0       138.0  56.601449  7.962082  35.0  58.0  77.0
1       165.0  52.496970  9.550651  29.0  52.0  76.0

In [25]:
## Ваш ответ:52.496970 i 9.550651

** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

> 0/1 баллов

In [6]:
(data[data['target']==0]['cholesterol'].mean())<(data['cholesterol'].mean())

False

In [66]:
x = data['cholesterol'].describe()
y = data[(data['sex'] == 'male') & (data['target'] == 0)]['cholesterol'].mean()
print(x)
print(y)

count    303.000000
mean     246.264026
std       51.830751
min      126.000000
25%      211.000000
50%      240.000000
75%      274.500000
max      564.000000
Name: cholesterol, dtype: float64
246.06140350877192


In [59]:
## Ваш ответ: Да. У них уровень хол. меньше:246.06140350877192 < 246.264026

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

> 1/1 баллов

In [32]:
data.groupby(['rest_ecg'])['age','resting_blood_pressure','cholesterol','max_heart_rate_achieved','st_depression','num_major_vessels','target'].agg([np.mean, np.max])

age      resting_blood_pressure       \
                                   mean amax                   mean amax   
rest_ecg                                                                   
ST-T wave abnormality         52.914474   71             129.065789  180   
left ventricular hypertrophy  61.000000   76             140.500000  180   
normal                        55.687075   77             134.027211  200   

                             cholesterol      max_heart_rate_achieved       \
                                    mean amax                    mean amax   
rest_ecg                                                                     
ST-T wave abnormality         237.269737  354              151.960526  194   
left ventricular hypertrophy  261.750000  327              125.750000  140   
normal                        255.142857  564              147.904762  202   

                             st_depression      num_major_vessels       \
                                      mean amax              mean amax   
rest_ecg                                                                 
ST-T wave abnormality             0.879605  5.6          0.638158    4   
left ventricular hypertrophy      2.725000  4.4          1.000000    3   
normal                            1.159184  6.2          0.816327    4   

                                target       
                                  mean amax  
rest_ecg                                     
ST-T wave abnormality         0.631579    1  
left ventricular hypertrophy  0.250000    1  
normal                        0.462585    1

In [35]:
## Ваш ответ: Таблица.

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

> 1/1 баллов

In [68]:
male = data[(data['sex'] == 'male') & (data['target'] == 1)]['st_depression'].mean()
female = data[(data['sex'] == 'female') & (data['target'] == 0)]['st_depression'].mean()
print(male)
print(female)

0.6053763440860215
1.8416666666666668


In [37]:
## Ваш ответ:У женщин выше.

**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

> 1/1 баллов

In [72]:
a = data.groupby(['chest_pain_type'])['cholesterol'].agg([np.min, np.max])
b = data.groupby(['rest_ecg'])['cholesterol'].agg([np.min, np.max])
c = data.groupby(['thalassemia'])['cholesterol'].agg([np.min, np.max])
print(a)
print(b)
print(c)

                  amin  amax
chest_pain_type             
atypical angina    126   564
non-anginal pain   182   298
typical angina     131   409
                              amin  amax
rest_ecg                                
ST-T wave abnormality          126   354
left ventricular hypertrophy   197   327
normal                         149   564
                   amin  amax
thalassemia                  
fixed defect        141   417
normal              169   318
reversable defect   126   564


In [ ]:
## Ваш ответ:Таблицы

**Сколько значений может принимать каждый из категориальных признаков?**

> 1/1 баллов

In [76]:
data.describe(include=[object])

sex chest_pain_type  fasting_blood_sugar               rest_ecg  \
count    303             303                  303                    303   
unique     2               3                    2                      3   
top     male  typical angina  lower than 120mg/ml  ST-T wave abnormality   
freq     207             193                  258                    152   

       exercise_induced_angina   st_slope   thalassemia  
count                      303        303           303  
unique                       2          2             3  
top                         no  upsloping  fixed defect  
freq                       204        161           166

In [ ]:
## Ваш ответ: Значение "unique" для каждого признака в таблице.

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

> 1/1 баллов

In [77]:
x = ['sex', 'chest_pain_type', 'fasting_blood_sugar', 'rest_ecg', 'exercise_induced_angina', 'st_slope', 'thalassemia']
w = data['sex'].value_counts(normalize=True)
l=1
p = 0
q = 0
    
for c in x:
    l=0
    z = data[c].value_counts(normalize=True)
    k = z[l]-z[l-1]
    if k>1*p:
        m = x[q]
        p = k
    q+=1
    for v in z:
        l+=1
print(p)
print(m)

0.7029702970297029
fasting_blood_sugar


In [ ]:
## Ваш ответ: ['fasting_blood_sugar']

### Молодец, 14/15

---

@Booster
Мамкин Data Scientist

tg @mommyscience